pkg/machines/hostvmslist.jsx

```diff
@@ -463,41 +467,93 @@ Vm.propTypes = {
 /**
  * List of all VMs defined on this host
  */
-const HostVmsList = ({ vms, config, dispatch, actions }) => {
-    if (vms.length === 0) {
-        return (<div className='container-fluid'>
-            <NoVm />
-        </div>);
+class HostVmsList extends React.Component {
+    constructor(props) {
+        super(props);
+        this.deviceProxyHandler = this.deviceProxyHandler.bind(this);
+        this.deviceProxyAddHandler = this.deviceProxyAddHandler.bind(this);
+        this.deviceProxyRemoveHandler = this.deviceProxyRemoveHandler.bind(this);
+        this.state = { hostDevices: []};
```

Comment:
I don't understand why you want to maintain this array? What don't you just expose client.proxies("org.freedesktop.NetworkManager.Device") in state. And look up the interface at cockpit.jump time?

Reply:
Because I think `VmNetworkTab` must get already prepared/pure data, it is easy to debug. But if for you it is important have proxy parsing in `VmNetworkTab`, I can move it.

Reply:
It makes sense to keep `vmNetworkTab` a presentational component (only properties, no state). However, it should work just as well to directly pass the NetworkManager proxy as a property to vmNetworkTab (what Peter suggested), instead of maintaining a copy here. This should also get rid of `deviceProxyAddHandler()` and friends.

Or is your concern not about presentational vs. stateful component, but about the ability to inject fake data into `hostDevices` for debugging? Then this would make sense, although it's quite expensive (seems better to plug in a complete fake object for the NM proxy then - you can even leave that code in for the future, if you comment it out).

Reply:
Ok, so I will make it as proxy.